src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java

```diff
@@ -0,0 +1,229 @@
+    }
+  }
+
+  public List<String> verifyAndGetInfoFromJwt(String assertion, SigningKeyResolverAdapter keys)
+      throws OAuth2Exception {
+    Jws<Claims> jws;
+
+    String email = null;
+    String googleAccountId = null;
+    try {
+      jws = Jwts.parserBuilder().setSigningKeyResolver(keys).build().parseClaimsJws(assertion);
+      if (!jws.getBody().getIssuer().equals("https://accounts.google.com")) {
+        throw new InvalidRequestException(ErrorCode.INVALID_JWT_ISS);
+      }
+      if (!jws.getBody().getAudience().equals(aud)) {
+        throw new InvalidRequestException(ErrorCode.WRONG_JWT_AUD);
+      }
+      email = jws.getBody().get("email", String.class);
+      googleAccountId = jws.getBody().get("sub", Long.class).toString();
+    } catch (JwtException ex) {
+      log.log(Level.INFO, "JWT Decode ERROR!", ex);
+      throw new InvalidRequestException(ErrorCode.INVALID_JWT);
+    }
+    return ImmutableList.of(email, googleAccountId);
+  }
+
+  public void handleCheckAssertion(
+      HttpServletResponse response, String email, String googleAccountId) throws IOException {
+    Optional<UserDetails> user =
+        userDetailsService.getUserByEmailOrGoogleAccountId(email, googleAccountId);
+    JSONObject json = new JSONObject();
+    if (user.isPresent()) {
+      json.appendField("account_found", "true");
+      response.setContentType("application/json;charset=UTF-8");
+      response.getWriter().println(json.toJSONString());
+    } else {
+      json.appendField("account_found", "false");
+      response.setContentType("application/json;charset=UTF-8");
+      response.getWriter().println(json.toJSONString());
+      response.setStatus(HttpStatus.SC_NOT_FOUND);
+    }
+    response.getWriter().flush();
+  }
+
+  public void handleGetAssertion(
+      HttpServletResponse response, String email, String googleAccountId, List<String> scopes)
+      throws IOException, InvalidScopeException {
+    Optional<UserDetails> user =
+        userDetailsService.getUserByEmailOrGoogleAccountId(email, googleAccountId);
+    if (user.isPresent()) {
+      ClientDetails client = clientDetailsService.getClientByID("google").get();
+
+      if (!scopes.isEmpty()
+          && client.getIsScoped()
+          && !client.getScopesList().containsAll(scopes)) {
+        throw new InvalidScopeException();
+      }
+
+      if (scopes.isEmpty() && client.getIsScoped()) {
+        scopes = client.getScopesList();
+      }
+      returnToken(response, client, user.get(), scopes);
+
+    } else {
+      returnLinkError(response, email);
+    }
+  }
+
+  public void handleCreateAssertion(
+      HttpServletResponse response, String email, String googleAccountId, List<String> scopes)
+      throws IOException, InvalidScopeException {
+    Optional<UserDetails> user =
+        userDetailsService.getUserByEmailOrGoogleAccountId(email, googleAccountId);
+    if (user.isPresent()) {
+      returnLinkError(response, email);
```

Comment:
Is there any place mentioned we shall return error when user exists?

Reply:
Yes.
Check whether either of the following conditions are true:

The Google Account ID, found in the assertion's sub field, is in your user database.
The email address in the assertion matches a user in your user database.
If either condition is true, prompt the user to link their existing account with their Google Account. To do so, respond to the request with an HTTP 401 error that specifies error=linking_error and gives the user's email address as the login_hint.

Reply:
SG